Rust code

ciaranra opened the pull request:
Including Rust code for stabilizer simulator.

TODO:
- [x] Have CI workflow that checks Rust tests and formatting
- [x] Update Makefiles to build etc. both Rust and Python projects given restructuring 
- [x] Have CI workflow that builds and install pecos-rslib and tests pecos-rslib
- [x] Have CI workflow that tests the usage of pecos-rslib from Python PECOS
- [x] Add PyO3 wrapping for Rust simulators and make available to Python via pecos-rslib
- [x] Update tests to test Rust simulators used in Python PECOS
- [x] Add RSSparseSim to Python `test_random_circuits` (likely need to add force_outcome measurement)
- [x] Add Rust tests for the newly added gates for the Rust stabilizer sim
- [x] Update/modify various Readmes as appropriate...
- [x] Use symlinks to point to the appropriate legal and Readme stuff
- [x] Have CI workflow that is generates sdists and wheels for all the Python packages
- [x] Add version consistency checks for Python packages and Rust crates
- [x] Update workflow to save artifacts only on success merges, direct pushes, or py- tags in main branches
- [x] Update README.md if needed

ciaranra commented:
TODO: Get workflow working again. Sort out makefiles.

ciaranra commented:
Workflow working again.
Current TODO:
- Have a workflow that checks Rust
- Get Makefiles sorted
- Have a workflow that builds binaries for pecos-rslib
- After Rust is tested & pecos-rslib binaries are built: then run Python tests

ciaranra commented:
In another PR I might consider switching to Justfile, Pixi, and/or Rye or other solutions to deal with a multi-language project + installing Cuda more seamlessly as long as complexity/user ease of use is managed/improved.

ciaranra commented:
Documentation is another thing that needs to be handled... but that is a large enough issue to be put off to another PR.

ciaranra commented:
Need to handle version bumping nicely... in another PR. Current thought is to have Rust and Python versioning independent but all projects of the same language have the same version number.

ciaranra commented:
Need to consider whether crates/pecos-pyo3 and python/pecos-rslib should just be one... Simplest first? 🤔

Do people want their own Rust PyO3 bindings without the Python side? ...

ciaranra commented:
Need more exhaustive testing for simulators. Maybe rather than forcing random outcomes... mirror benchmarking... Although maybe that assumes too much.

Need to go through everything and standardize gate symbols... and how to call sims...

peter-campora commented:
Oh boy CRA, this is a monster PR. I think it would help if you have some examples of the kinds of tests you would like the PR reviewers to run on this

ciaranra commented:
Hehe, yeah it is a monster. Let me single out some smaller stuff for people to maybe concentrate on if they want to...

ciaranra commented:
If you are up for reviewing this monster of a PR, here are some things that you can take a look at:

- Makefile: After reading the README and checking out the Makefile is it clear what to do? Do the relevant targets make sense? Maybe take the makefile out for a spin and see if you can build and test with it?
- README: Do you think there is anything glaringly missing for users or developers?
-  Workflows: New workflows have been added to test and build both the Rust and Python code: `.github/workflows`. If you want you can see if all the testing and linting look right. The workflows look like they are doing their jobs properly... `python-release.yml` is a bit of a beast, but it is mainly a utility for building distributions to be uploaded to pypi
- tests... the tests are fairly extensive and probably more for people who are comfortable with stabilizer simulations. If you want to take a look at them... they are in `crates/pecos-qsims/src/sparse_stab.rs` (at the bottom of the file), in here `python/tests/pecos/integration/state_sim_tests/test_stab_sims/`, and here `python/tests/pecos/integration/test_random_circuits.py`. The Rust stabilizer simulator is probably better tested then the Python one. lol
- General structure: Does the general approach and structure of files for this monolithic repo seem okay?

ciaranra commented:
New TODO:
- [ ] Add license headers to all code

qartik commented on README.md:
````diff
@@ -42,64 +58,75 @@ Stay updated with the latest developments on the
+
+3. Run tests:
+   ```sh
+   make test
````
I get
```
❯ make test -n
make: *** No rule to make target `test'.  Stop.
```

ciaranra replied:
Changed this to `make pytest`

qartik commented on README.md:
````diff
@@ -42,64 +58,75 @@ Stay updated with the latest developments on the
+
+4. Run pre-commit (after [installing it](https://pre-commit.com/))
+   ```sh
+   make pre-commit
````
I wish there was a single step to set up the virtual environment and not all make commands be calling `make venv`, which can be slow and annoying.

ciaranra replied:
The `venv` target has been removed from others.

qartik reviewed (commented):
Mostly just tried to work through the development section of the README.

`make test-all` gave me:
```
======== short test summary info ========
FAILED docs/api_guide/circuit_runners.rst::circuit_runners.rst
FAILED docs/api_guide/decoders.rst::decoders.rst
FAILED docs/api_guide/error_generators.rst::error_generators.rst
FAILED docs/api_guide/qeccs.rst::qeccs.rst
FAILED docs/api_guide/simulators.rst::simulators.rst
FAILED docs/examples/creating_qecc_class.rst::creating_qecc_class.rst
FAILED docs/examples/stab_code_verification.rst::stab_code_verification.rst
======== 7 failed, 2 passed, 1 warning in 0.82s ========
```

peter-campora commented on crates/benchmarks/benches/modules/helpers/rot_surface.rs:
```diff
@@ -0,0 +1,276 @@
+    distance: usize,
+    width: usize,
+    height: usize,
+    data_qubit_ids: Vec<usize>,
```
Nit: What are the usizes supposed to represent in this type? Are they related to the width and height in the `RotSurface` struct? It may help to either use a named struct or in the case that the tuple is easier to use, oruse type aliases for the different usizes so the interpretation is easy to grok from reading the type. Or alternately document what the type means in a comment. Also, I noticed that though the `layout` field in `RotSurface` is the same type, this type alias is not used.

ciaranra replied:
This is connected with pecos-qec not being ready yet. Will fix.

peter-campora commented on Makefile:
```diff
@@ -1,102 +1,170 @@
+# Building development environments
+# ---------------------------------
+.PHONY: build
+build: venv ## Compile and install for development
```
This fails for me with: 
```bash
RROR: Could not find a version that satisfies the requirement pecos-rslib==0.6.0.dev6 (from quantum-pecos) (from versions: none)
ERROR: No matching distribution found for pecos-rslib==0.6.0.dev6
```

I am assuming I need to make requirements first? I will try that and then retry `make build`

peter-campora replied:
Hmmm same error, but looking at the fuller dump, I'm very confused because it looks like the package is being installed and then not used:

```bash
📦 Including license file "/home/peter/Downloads/PECOS-feat-rust/python/pecos-rslib/LICENSE"
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings
🐍 Found CPython 3.10 at /home/peter/Downloads/rust-pecos/bin/python
📡 Using build options features from pyproject.toml
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.07s
📦 Built wheel for CPython 3.10 to /tmp/.tmpzQTl39/pecos_rslib-0.6.0.dev6-cp310-cp310-linux_x86_64.whl
✏️  Setting installed package as editable
🛠 Installed pecos-rslib-0.6.0.dev6
Looking in indexes: https://john.campora%40quantinuum.com:****@quantinuumhw.jfrog.io/artifactory/api/pypi/qtm-pypi-vc/simple
Obtaining file:///home/peter/Downloads/PECOS-feat-rust/python/quantum-pecos
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of quantum-pecos to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement pecos-rslib==0.6.0.dev6 (from quantum-pecos) (from versions: none)
ERROR: No matching distribution found for pecos-rslib==0.6.0.dev6
make: *** [Makefile:75: build] Error 1
```

ciaranra replied:
Okay, now you will have to make a .venv either via the `venv` command or by hand. If this doesn't "fix" this. Might need to take out all the $(VENV_BIN)/ and assume the user has dropped into the virtual environment

peter-campora commented on README.md:
````diff
@@ -42,64 +58,75 @@ Stay updated with the latest developments on the
-### Simulators with special requirements
+2. Set up the development environment using the makefile (for Windows you may need to use a shell that has access to Linux commands such as utilizing [git bash](https://gitforwindows.org/)):
+   ```sh
+   make venv
+   make build
````
Fuller dump:

```bash
📦 Including license file "/home/peter/Downloads/PECOS-feat-rust/python/pecos-rslib/LICENSE"
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings
🐍 Found CPython 3.10 at /home/peter/Downloads/rust-pecos/bin/python
📡 Using build options features from pyproject.toml
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.07s
📦 Built wheel for CPython 3.10 to /tmp/.tmpzQTl39/pecos_rslib-0.6.0.dev6-cp310-cp310-linux_x86_64.whl
✏️  Setting installed package as editable
🛠 Installed pecos-rslib-0.6.0.dev6
Looking in indexes: https://john.campora%40quantinuum.com:****@quantinuumhw.jfrog.io/artifactory/api/pypi/qtm-pypi-vc/simple
Obtaining file:///home/peter/Downloads/PECOS-feat-rust/python/quantum-pecos
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of quantum-pecos to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement pecos-rslib==0.6.0.dev6 (from quantum-pecos) (from versions: none)
ERROR: No matching distribution found for pecos-rslib==0.6.0.dev6
make: *** [Makefile:75: build] Error 1

```

peter-campora commented on Makefile:
```diff
@@ -1,102 +1,170 @@
-clean:  ## Clean up caches and build artifacts
-	-rm -rf *.egg-info dist build docs/_build .pytest_cache/ .ruff_cache/
+.PHONY: test-all
+test-all: rstest pytest pytest-dep pydoctest ## Run all tests. ASSUMES: previous build command
```
I'm getting a lot errors of this `pecos is not a package` form: 

```bash
___________________________________________ ERROR collecting pecos/unit/test_blocks.py ____________________________________________
ImportError while importing test module '/home/peter/Downloads/PECOS-feat-rust/python/tests/pecos/unit/test_blocks.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
pecos/unit/test_blocks.py:13: in <module>
    from pecos.classical_interpreters.phir_classical_interpreter import (
E   ModuleNotFoundError: No module named 'pecos.classical_interpreters'; 'pecos' is not a package
```

Full test failure list:
```bash
ERROR pecos/integration/example_tests/test_basic_logical_sim.py
ERROR pecos/integration/example_tests/test_finding_threshold.py
ERROR pecos/integration/example_tests/test_recovery.py
ERROR pecos/integration/state_sim_tests/test_cointoss.py
ERROR pecos/integration/state_sim_tests/test_stab_sims/test_gate_init.py
ERROR pecos/integration/state_sim_tests/test_stab_sims/test_gate_one_qubit.py
ERROR pecos/integration/state_sim_tests/test_stab_sims/test_gate_two_qubit.py
ERROR pecos/integration/state_sim_tests/test_statevec.py
ERROR pecos/integration/test_phir.py
ERROR pecos/integration/test_phir_64_bit.py
ERROR pecos/integration/test_phir_setting_cregs.py
ERROR pecos/integration/test_quantum_circuits.py
ERROR pecos/integration/test_random_circuits.py
ERROR pecos/regression/test_qasm/examples/test_logical_steane_code_program.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_measures.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_preps.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_rots.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_sq_face_rots.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_sq_hadamards.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_sq_noncliffords.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_sq_paulis.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_sq_sqrt_paulis.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_tq_cliffords.py
ERROR pecos/regression/test_qasm/pecos/qeclib/qubit/test_tq_noncliffords.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/decoders/test_lookup.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/gates_sq/test_face_rots.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/gates_sq/test_hadamards.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/gates_sq/test_paulis.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/gates_sq/test_sqrt_paulis.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/gates_tq/test_transversal_tq.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/meas/test_destructive_meas.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/meas/test_measure_x.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/meas/test_measure_z.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/preps/test_encoding_circ.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/preps/test_pauli_states.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/preps/test_plus_h_state.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/preps/test_t_plus_state.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/qec/test_qec_3parallel.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/syn_extract/test_six_check_nonflagging.py
ERROR pecos/regression/test_qasm/pecos/qeclib/steane/syn_extract/test_three_parallel_flagging.py
ERROR pecos/regression/test_qasm/random_cases/test_control_flow.py
ERROR pecos/regression/test_qasm/random_cases/test_slr_phys.py
ERROR pecos/unit/reps/pymir/test_name_resolver.py
ERROR pecos/unit/test_binarray.py
ERROR pecos/unit/test_blocks.py
```

ciaranra replied:
Did you follow this note? ## Run all tests. ASSUMES: previous build command

Although, if you are having issues... probably need to find another solution. :D

ciaranra replied:
Weird. Try again please 🙂

peter-campora commented on crates/pecos-core/src/sims_rngs/mock_rng.rs:
```diff
@@ -0,0 +1,91 @@
+        Self { values, index: 0 }
+    }
+
+    // pub fn gen_bool(&mut self, _p: f64) -> bool {
```
Delete dead code?

peter-campora commented on crates/pecos-core/src/sims_rngs/sim_rng.rs:
```diff
@@ -0,0 +1,64 @@
+        SeedableRng::from_entropy()
+    }
+
+    // fn seed_from_u64(seed: u64);
```
Delete dead bindings?

peter-campora commented on crates/pecos-qec/src/rot_surface/patch_layout.rs:
```diff
@@ -0,0 +1,89 @@
+
+#[allow(clippy::module_name_repetitions)]
+#[must_use]
+pub fn gen_patch_layout(height: usize, width: usize) -> RotSurfLayout {
```
Nit: the way the mutable references are passed to `new_node` seems like it could be refactored but probably not worth the hassle unless you start getting into borrow checker issues

ciaranra replied:
I think there is some Rust code that just needs to be removed for this release that I left in that isn't fully baked yet. I'll clean it up and wait to introduce it later on. Need another dev branch for my dev branch.

ciaranra replied:
Basically, ignore `pecos-qec` for now.

peter-campora commented on crates/pecos-qsims/src/pauli_prop.rs:
```diff
@@ -0,0 +1,223 @@
+    E: IndexableElement,
+{
+    #[inline]
+    fn px(&mut self, q: E) -> (bool, bool) {
```
If you have a reference on how these functions should be implemented, then go you should consider making an issue with the right docs and maybe you can get some help filling them out

ciaranra replied:
Issue created: #83
We'll see how much energy I have to deal with it now... but given the size of this PR. Maybe I should give GitHub a breather. :D

ciaranra commented:
> Mostly just tried to work through the development section of the README.
> 
> `make test-all` gave me:
> 
> ```
> ======== short test summary info ========
> FAILED docs/api_guide/circuit_runners.rst::circuit_runners.rst
> FAILED docs/api_guide/decoders.rst::decoders.rst
> FAILED docs/api_guide/error_generators.rst::error_generators.rst
> FAILED docs/api_guide/qeccs.rst::qeccs.rst
> FAILED docs/api_guide/simulators.rst::simulators.rst
> FAILED docs/examples/creating_qecc_class.rst::creating_qecc_class.rst
> FAILED docs/examples/stab_code_verification.rst::stab_code_verification.rst
> ======== 7 failed, 2 passed, 1 warning in 0.82s ========
> ```

This should be resolved now. Just removing doc tests for now.

ciaranra commented:
Alright, @qartik & @peter-campora. I have updated the makefile to make the .venv more of a "first class citizen" and updated the README appropriately (hopefully). Let me know if you still have issues.

I still need to figure out what is going on with this: 
> -pecos-rslib==0.6.0.dev6
> +pecos-rslib==version = 0.6.0.dev6

ciaranra commented on Makefile:
```diff
@@ -1,102 +1,170 @@
-metadeps: upgrade-pip  ## Install extra dependencies used to develop/build this package
-	$(VENV_BIN)/pip install -U build pip-tools pre-commit wheel pytest hypothesis
+.PHONY: venv
+venv:  ## Build Python virtual environment and install requirements
```
`make venv` no longer calls `make updatereqs`. Running `make updatereqs`, I get `pecos-rslib==0.6.0dev6` and not the other one. I am wondering if this is a Mac subtlety.

Can you reproduce this behavior now? And if so, what are the steps?
